script/generate_async.py

```diff
@@ -0,0 +1,105 @@
+"""
+Script to automatically generate grafana asynchronous code from the
+synchronous counterpart.
+What does this program do:
+- For all module in grafana_client/elements/*.py excepted base and __init__
+Inject async/await keyword in all available methods / client interactions.
+- Then detect no longer needed module for removal.
+- Finally generate the _async top level code based on the elements/__init__.py one.
+"""
+
+import os
+import re
+import subprocess
+from glob import glob
+
+BASE_PATH = ".." if os.path.exists("../grafana_client") else "."
+
+if __name__ == "__main__":
+    module_processed = []
+    module_generated = []
+
+    for module_path in glob(f"{BASE_PATH}/grafana_client/elements/*.py"):
+        if "__init__.py" in module_path or "base.py" in module_path:
+            continue
+
+        print(f"Processing {module_path}...")
+
+        with open(module_path, "r") as fp:
+            module_dump = fp.read()
+
+        module_dump = re.sub(r"( {4}def )(?!_)", r"    async def ", module_dump)
+
+        module_dump = module_dump.replace("self.client.", "await self.client.")
+
+        for relative_import in [".base", "..client", "..knowledge", "..model"]:
+            module_dump = module_dump.replace(f"from {relative_import}", f"from .{relative_import}")
+
+        module_dump = module_dump.replace("self.api.version", "await self.api.version")
```

Comment:
Now that I am seeing it, it is clear, but still let me have a question about it: In Python/async, the caller will need to know that some synthetic attribute, implemented as a Python `@property`, is actually an `async` method, and invoke it accordingly?

I mean it's somehow clear, but still violates some principle of abstraction / decoupling, when being overly nitpicky about it, at least semantically, no?

This is also really just a question, and nothing to address. Thanks!

Reply:
I understand the question on the principles.
Unfortunately I don't have the knowledge on "the why" it was implemented this way by Python. But interesting to dig sometime.